File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    scores = []
    temp_num = 0
    
    # 문자열을 하나씩 순회
    for i in range(len(dartResult)):
        char = dartResult[i]
        
        # 1. 숫자 처리 (0~10)
        if char.isdigit():
            # 만약 10인 경우 (앞글자가 1이고 현재가 0일 때)
            if char == '0' and i > 0 and dartResult[i-1] == '1':
                temp_num = 10
            else:
                temp_num = int(char)
                
        # 2. 보너스 처리 (S, D, T)
        elif char == 'S':
            scores.append(temp_num ** 1)
        elif char == 'D':
            scores.append(temp_num ** 2)
        elif char == 'T':
            scores.append(temp_num ** 3)
            
        # 3. 옵션 처리 (*, #)
        elif char == '*':
            # 스타상: 현재 점수 2배
            scores[-1] *= 2
            # 이전 점수가 있다면 그것도 2배
            if len(scores) >= 2:
                scores[-2] *= 2
        elif char == '#':
            # 아차상: 현재 점수 마이너스
            scores[-1] *= -1
            
    return sum(scores)

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.js
function solution(dartResult) {
    const result = [];
    let temp_num = 0;

    for (let i = 0; i < dartResult.length; i++) {
        const char = dartResult[i];

        // 1. 숫자 처리
        if (char >= "0" && char <= "9") {
            // 만약 숫자가 두자리인경우
            if (char === "0" && i > 0 && dartResult[i - 1] === "1") {
                temp_num = 10;
            } else {
                temp_num = parseInt(char);
            }

            // 2. 보너스 처리
        } else if (char === "S") {
            result.push(temp_num ** 1);
        } else if (char === "D") {
            result.push(temp_num ** 2);
        } else if (char === "T") {
            result.push(temp_num ** 3);
        }

        // 3. 옵션 처리
        else if (char === "*") {
            // 스타상: 현재 점수 2배
            result[result.length - 1] *= 2;
            if (result.length >= 2) {
                result[result.length - 2] *= 2;
            }
        } else if (char === "#") {
            // 아차상: 현재 점수 마이너스
            result[result.length - 1] *= -1;
        }
    }
    return result.reduce((acc, cur) => acc + cur, 0);
}
